go-controller/pkg/ovn/service.go

```diff
@@ -285,11 +286,34 @@ func (ovn *Controller) createService(service *kapi.Service) error {
 }
 
 func (ovn *Controller) updateService(oldSvc, newSvc *kapi.Service) error {
-	klog.V(5).Infof("Updating service is a noop: %s", newSvc.Name)
-	// Service update needs to check for port change, protocol, etc and update the ACLs, or and OVN LBs
-	// This only really matters when a service is updated and has no endpoints. If the service has endpoints
-	// the endpoints watcher will handle the OVN config
-	// TODO (trozet) implement this
+	if reflect.DeepEqual(newSvc.Spec.Ports, oldSvc.Spec.Ports) && reflect.DeepEqual(newSvc.Spec.ExternalIPs, oldSvc.Spec.ExternalIPs) {
+		klog.V(5).Infof("skipping service update as change does not apply to .Spec.Ports or .Spec.ExternalIP")
+		return nil
+	}
+
+	if !util.IsClusterIPSet(newSvc) && !util.IsClusterIPSet(oldSvc) {
+		klog.V(5).Infof("skipping service update due to service being headless")
+		return nil
+	}
+
+	klog.V(5).Infof("updating service from: %v to: %v", oldSvc, newSvc)
+
+	// New service or old service, it's all the same for what concerns the endpoint.
+	ep, err := ovn.watchFactory.GetEndpoint(newSvc.Namespace, newSvc.Name)
+	if err != nil {
+		klog.V(5).Infof("updating service with no endpoint, service: %s, namespace: %s", newSvc.Name, newSvc.Namespace)
+	}
+
+	ovn.deleteService(oldSvc)
```

Comment:
I know typically in our update of resources we just do a delete/re-add. This is a simple thing to do, but it's not necessarily the best thing to do. Service is a prime example of this, where we could just be updating a load balancer VIP. In that case it isn't great that we are entirely clearing the load balancer in OVN and bringing traffic down, instead of just modifying the load balancer appropriately (configureLoadBalancer).

Reply:
I agree, and that's what I started doing. The problem we face is that technically everything is allowed when it comes to service updates, the amount of ports, names of ports, port number, etc. This makes tracking the changes and updating OVN correctly very complicated, and the code I wrote started taking proportions which made me believe it would add more bugs than it would solve. So let's just do it like this for now until we invent a proper mechanism for detecting changes and acting accordingly.